File: frontend/src/lib/types/recipe.ts
/**
 * Recipe-related types
 */

export type Difficulty = 'easy' | 'medium' | 'hard';
export type SourceType = 'manual' | 'scraped';

export interface Ingredient {
  id: string;
  name: string;
  quantity?: number;
  unit?: string;
  category?: string;
  is_optional?: boolean;
}

export interface IngredientInput {
  name: string;
  quantity?: number;
  unit?: string;
  category?: string;
  is_optional?: boolean;
}

export interface RecipeSummary {
  id: string;
  title: string;
  description?: string;
  prep_time_minutes?: number;
  cook_time_minutes?: number;
  servings?: number;
  difficulty?: Difficulty;
  tags?: string[];
  is_favorite?: boolean;
  last_cooked_date?: string;
}

export interface Recipe extends RecipeSummary {
  source_url?: string;
  source_type?: SourceType;
  current_version: number;
  ingredients: Ingredient[];
  instructions: string;
  images?: string[];
  nutritional_info?: Record<string, unknown>;
  created_by: string;
  created_at: string;
}

export interface RecipeCreate {
  title: string;
  description?: string;
  prep_time_minutes?: number;
  cook_time_minutes?: number;
  servings?: number;
  difficulty?: Difficulty;
  ingredients: IngredientInput[];
  instructions: string;
  tags?: string[];
  source_url?: string;
}

export interface RecipeUpdate extends RecipeCreate {
  change_description?: string;
}

export interface RecipeVersion {
  version_number: number;
  prep_time_minutes?: number;
  cook_time_minutes?: number;
  servings?: number;
  difficulty?: Difficulty;
  ingredients: Ingredient[];
  instructions: string;
  change_description?: string;
  modified_by: string;
  created_at: string;
}

export interface RecipeVersionsResponse {
  recipe_id: string;
  current_version: number;
  versions: RecipeVersion[];
}

export interface ScrapeRecipeRequest {
  url: string;
}

export interface ScrapeRecipeResponse {
  scraped_data: RecipeCreate;
  source_url: string;
  warnings?: string[];
}

export interface RecipeSuggestion {
  recipe: RecipeSummary;
  score: number;
  reasons: string[];
}

export interface RecipeSuggestionsResponse {
  suggestions: RecipeSuggestion[];
}

export type RecipeFilter = 'favorites' | 'not_recent' | 'never_tried' | 'available_inventory';

File: backend/src/schemas/recipe.py
"""
Recipe Schemas
Pydantic models for recipe-related operations
"""

from datetime import date, datetime
from decimal import Decimal
from typing import Any, Dict, List, Optional
from uuid import UUID

from pydantic import BaseModel, Field, field_validator


class IngredientInput(BaseModel):
    """Schema for creating/updating ingredients"""

    name: str = Field(..., description="Ingredient name")
    quantity: Optional[Decimal] = Field(None, ge=0, description="Quantity")
    unit: Optional[str] = Field(None, description="Unit of measurement")
    category: Optional[str] = Field(None, description="Ingredient category")
    is_optional: Optional[bool] = Field(False, description="Optional ingredient flag")


class IngredientResponse(IngredientInput):
    """Schema for ingredient response"""

    id: UUID
    display_order: int

    class Config:
        from_attributes = True


class RecipeBase(BaseModel):
    """Base recipe schema"""

    title: str = Field(..., max_length=255, description="Recipe title")
    description: Optional[str] = Field(None, description="Recipe description")
    prep_time_minutes: Optional[int] = Field(
        None, ge=0, description="Preparation time in minutes"
    )
    cook_time_minutes: Optional[int] = Field(
        None, ge=0, description="Cooking time in minutes"
    )
    servings: Optional[int] = Field(None, gt=0, description="Number of servings")
    difficulty: Optional[str] = Field(None, description="Difficulty level")
    tags: Optional[List[str]] = Field(default_factory=list, description="Recipe tags")
    source_url: Optional[str] = Field(None, description="Source URL")

    @field_validator("difficulty")
    @classmethod
    def validate_difficulty(cls, v):
        if v is not None and v not in ["easy", "medium", "hard"]:
            raise ValueError("Difficulty must be easy, medium, or hard")
        return v


class RecipeCreate(RecipeBase):
    """Schema for creating a recipe"""

    ingredients: List[IngredientInput] = Field(
        ..., min_length=1, description="Recipe ingredients"
    )
    instructions: str = Field(..., min_length=1, description="Cooking instructions")


class RecipeUpdate(RecipeCreate):
    """Schema for updating a recipe (creates new version)"""

    change_description: Optional[str] = Field(
        None, description="Description of changes"
    )


class RecipeSummary(BaseModel):
    """Schema for recipe summary (list view)"""

    id: UUID
    title: str
    description: Optional[str] = None
    prep_time_minutes: Optional[int] = None
    cook_time_minutes: Optional[int] = None
    servings: Optional[int] = None
    difficulty: Optional[str] = None
    tags: List[str] = Field(default_factory=list)
    is_favorite: bool = False
    last_cooked_date: Optional[date] = None

    class Config:
        from_attributes = True


class RecipeResponse(RecipeSummary):
    """Schema for detailed recipe response"""

    source_url: Optional[str] = None
    source_type: str
    current_version: int
    ingredients: List[IngredientResponse]
    instructions: str
    images: List[str] = Field(default_factory=list)
    nutritional_info: Optional[Dict[str, Any]] = None
    created_by: UUID
    created_at: datetime

    class Config:
        from_attributes = True


class RecipeVersionResponse(BaseModel):
    """Schema for recipe version response"""

    version_number: int
    prep_time_minutes: Optional[int] = None
    cook_time_minutes: Optional[int] = None
    servings: Optional[int] = None
    difficulty: Optional[str] = None
    ingredients: List[IngredientResponse]
    instructions: str
    change_description: Optional[str] = None
    modified_by: UUID
    created_at: datetime

    class Config:
        from_attributes = True


class ScrapedRecipeResponse(BaseModel):
    """Schema for scraped recipe response"""

    scraped_data: RecipeCreate
    source_url: str
    warnings: List[str] = Field(default_factory=list)
